src/black/__init__.py

```diff
@@ -3287,8 +3287,11 @@ def do_match(self, line: Line) -> TMatchResult:
                 and LL[next_idx].type == token.RPAR
                 and not is_empty_rpar(LL[next_idx])
             ):
-                # That RPAR should NOT be followed by a '.' symbol.
-                if is_valid_index(next_idx + 1) and LL[next_idx + 1].type == token.DOT:
+                # That RPAR should NOT be followed by a '.' or a '['
+                if is_valid_index(next_idx + 1) and LL[next_idx + 1].type in {
+                    token.DOT,
+                    token.LSQB,
```

Comment:
These similar cases still crash:
- `echo '("x" % 3)(1)' | black -`.
- `echo '("x" % 3) ** 1' | black -`.
Could you fix these too?

(Essentially, we should skip anything lower in https://docs.python.org/3/reference/expressions.html#operator-precedence than % is.)

Reply:
sure. i don't quite understand why only `%` has this problem. (e.g. why does `'("x" or 3)(1)'` work fine?)

Reply:
ah, there are more issues. operators in the same group are ok after, because it works left to right, but e.g. `'2*("" % a)` still breaks. investigating

Reply:
ok, added another commit for that

Reply:
hm, this might be quite a strong restriction, other stuff is failing now.  maybe we need to detect the percent sign and only conditionally skip rules?  thoughts?

Reply:
Looks like the reason only % is affected is the code in the StringParser class that this class relies on to decide how big the string is. The code looks right to me now and has thorough tests, so I'll merge the PR. Thanks for your contribution!